### ISSUE TITLE
[BUG] "Auto-indent" option leaves blank lines indented

### ISSUE THREAD

**Issue by aaaxx** [notepad-plus-plus/notepad-plus-plus#443]
Very annoying behaviour that's not present in other editors.
After hitting Enter at the end of an indented line, the next line is, of course, automatically indented to the same level as the previous one. However, if you hit Enter again, leaving the line empty, the indent will stay instead of being trimmed automatically.
```
→ → code
→ →
→ → code
```

**Comment by h-h-h-h** [comment]
I guess the first behavior that you describe is supposed to be this way. If a developer wants to have a blank line in indented code, he would desire exactly this. When working with lists it may be not desired. So make it an option?

For the second behavior that you describe I can confirm that this can be annoying.

**Comment by bezz92** [comment]
As a workaround for this you may use the macro "alt + shift + s" (available by default in a notepad++ installation). This will trim the extraneous white space you encounter when using auto-indent and will then save your document.

**Comment by tvb377** [comment]
Would an option to disable this behaviour be desirable / could it be agreed upon?
If so I'd try to implement this.

**Comment by h-h-h-h** [comment]
> @tvb377:
Would an option to disable this behaviour be desirable / could it be agreed upon?

If others dislike this, there could be an option for it, although *I* wouldn't need it:

> After hitting Enter at the end of an indented line, the next line is, of course, automatically indented to the same level as the previous one. However, if you hit Enter again, leaving the line empty, the indent will stay instead of being trimmed automatically.

This should generally be changed:

> ```
→ <Enter>
<Backspace 1x> <Enter>
→
```

**Comment by dg9ngf** [comment]
The last point is probably the bug that is by far most annoying me today. I use indented text files a lot to take notes. When I add a blank line in front of an unindented line, that happens to be after a line indented by one tab, that previously unindented line now becomes indented! Very much not expected. I hope this issue will be the next thing being worked on, if anything is done at NPP at all (I don't see many buges fixed, just more new ones opened).

**Comment by ygoe** [comment]
Any news on this? It keeps bugging me every single day I work with it. Does this project have any kind of priority of bugs over features? I'd rather have a tool that does well what it does, instead of one that does everything just mean. There have been at least two new feature releases alone since the last comment here. I don't have much time to work on this tool besides the other tools I'm already making for my work, but is there any news on building Notepad++ with Visual Studio 2015 so I could at least try to take a look at this bug?

**Comment by A-R-C-A** [comment]
> ...is there any news on building Notepad++ with Visual Studio 2015 so I could at least try to take a look at this bug?

NotepadPlus.vs2015.vcxproj is included, so you can start fixing.
Look here: `void Notepad_plus::maintainIndentation(TCHAR ch)` in `Notepad_plus.cpp`.

**Comment by ygoe** [comment]
@A-R-C-A There is still the outdated Boost library that requires VS 2013. Could somebody please fix that first? I can't do anything otherwise. I've tried it before and failed.

**Comment by MAPJe71** [comment]
For building Scintilla with Boost > v1.55 and using VS2015 see PR #765

**Comment by MAPJe71** [comment]
@ygoe Don't feel obligated to make any patches just because I pointed out the PR.
I remembered someone had made a patch because I made a similar patch before that (never created a PR for it) as I ran into the same problem with trying to use Boost 1.56 for my FunctionList patch (nr. 548 on SourceForge).

**Comment by A-R-C-A** [comment]
@ygoe
You can focus on building Npp (PowerEditor) only. There is no need to build scintilla, just copy an existing `SciLexer.dll` (e.g. from official release) to your testing folder.

Alternatively you can build scintilla without Boost, see the `README.md` for more information.

**Comment by ygoe** [comment]
As opposed to the initial report, the indenting is wrongly restored for any level.

    → → → Text<Enter>
    <Backspace 3x><Enter>
    → → → 

It also happens when you're not typing all this straight ahead but insert a new line somewhere after an indented line.

I'd expect that auto-indenting should only happen when I press Enter at the end of a line. Not when I press Enter at the beginning of the following lines. **And certainly not when that line is not empty,** because that will not only move that line down but indent it and thereby modify its meaning.

And it is annoying every single day. I'm constantly fixing up the mess Notepad++ causes.

Does this bug get any attention by someone who is able to handle the code? Maybe even before we get more shiny features? I understand that it is more interesting to create new features, but technical debt will catch you some day and it is by far less appealing to users who are actually affected by this. And I guess that's > 50 %. Annoyed users just won't appreciate your new features.

**Comment by dcomtois** [comment]
See this link for a plugin that helps:

http://www.virtualroadside.com/blog/index.php/2012/12/03/notepad-plugin-to-enable-unindent-on-backspace-key/

**Comment by ygoe** [comment]
@dcomtois According to its description, it only affects backspace-unindenting with spaces instead of tabs. (I'm using tabs here and most elsewhere.) It should have no effect on automatic indenting on line breaks.

**Comment by dcomtois** [comment]
@ygoe Sorry about the confusion. I deleted my comment.

**Comment by Beep6581** [comment]
The annoying behavior where indented empty lines appear is still present in Notepad++ 7.5.4.

Current behavior:
https://filebin.net/1dow2qwm7v7o3vpg/vid1-2018-08-01_17.02.19.mp4?t=oh37zfkf
Notice how empty lines are till filled with space characters.

Expected behavior (vim-8.1.0034)
https://filebin.net/1dow2qwm7v7o3vpg/vid1-2018-08-01_17.04.34.mp4?t=oh37zfkf
Notice how empty lines really are empty, but you can still type from the indent if you want to.

**Comment by vertigo220** [comment]
I would love for this issue to be fixed. It's one of the most common annoyances I deal with when using NPP. I think I'm going to just disable auto-indent most of the time, though it would be nice to be able to customize the toolbar to add a button to toggle it.

**Comment by cganchev** [comment]
I can't believe that this issue is still not fixed.
**Notepad2** is also based on **Scintilla** and handles auto-indent correctly.
The issue in **Notepad++** happens when you are at the **beginning of the line** and you press **Enter** - it should add a **blank line** above and **not change the indent of the current line**, but it changes the indent to match the one of the **first non-empty line above**.
It should be very easy to fix and Scintilla already supports this behavior.

Don't get me wrong, I like Notepad++ a lot and I use it daily, in fact for me that's the only annoying thing about it.

**Comment by ygoe** [comment]
If the Notepad++ source code isn't completely messed up (I'm not sure), it is easy to fix that.

It's just not easy to build it from source (with a current set of tools), or I'd already have fixed it myself. This thing keeps up my dreams of creating my own Scintilla-based editor. These dreams would die a little if Notepad++ would stop adding features and start fixing bugs.

**Comment by cganchev** [comment]
Thanks to @A-R-C-A for pointing the function which maintains the indentation:
`void Notepad_plus::maintainIndentation(TCHAR ch)` in `Notepad_plus.cpp`

I reviewed the code and the problem is in this function indeed.

It's easy to spot that it looks for a **non-empty previous line**:

```
			// Search the non-empty previous line
			while (prevLine >= 0 && _pEditView->getLineLength(prevLine) == 0)
				prevLine--;
```

It doesn't care what is the current cursor position.
The fix is quite easy, get the current position by calling `_pEditView->execute(SCI_GETCURRENTPOS)` and if it matches the beginning of the line (not sure if it's 0 or 1) then do nothing.
I don't know any other editor which messes with the indentation when you're in the beginning of the line and you press Enter.
Initially I thought that this is something managed by Scintilla but now I see that this is completely custom Notepad++ code, which explains why no other editor does the same.

**Comment by sasumner** [comment]
@cganchev said:

>The fix is quite easy...

So you should prepare a PR...

**Comment by cganchev** [comment]
I just created a PR which should fix the issue, but I haven't tested it.
Further optimizations are possible because some code is repeated in multiple places, but I decided to do it with minimal changes.

I didn't use `_pEditView->execute(SCI_GETCURRENTPOS)` because it returns the position from the beginning of the document, not in the current line. To determine if we were at the beginning of the line I am checking the length of the previous line. Probably this change is making the search for non-empty previous line redundant.

I hope that @donho will find some time to review the PR.

**Comment by sasumner** [comment]
@cganchev said:

>but I haven't tested it.

Nobody wants an untested PR.  :-(

**Comment by cganchev** [comment]
That's true, but the change is minimal and I can't afford to spend a lot of time to create environment where I can build Notepad++. Also previous messages state that it's not easy to build it anyway.
